File: medium/3016_minimum_number_of_pushes_to_type_word_ii/solution.cpp
#include <queue>
#include <string>
#include <unordered_map>

class Solution {
public:
  int minimumPushes(std::string word) {
    std::unordered_map<char, int> charsCounter;
    for (const char &ch : word)
      ++charsCounter[ch];

    std::priority_queue<int> pq;
    for (const auto &[_, count] : charsCounter)
      pq.push(count);

    int pushesCount = 0;
    int keysCount = 0;

    while (!pq.empty()) {
      pushesCount += ((keysCount++) / 8 + 1) * pq.top();
      pq.pop();
    }

    return pushesCount;
  }
};

File: medium/3016_minimum_number_of_pushes_to_type_word_ii/solution.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        charsCounter = [0] * 26
        for ch in word:
            charsCounter[ord(ch) - 97] += 1

        pushesCount = 0
        keysCount = 0

        for count in sorted(charsCounter, reverse=True):
            pushesCount += (keysCount // 8 + 1) * count
            keysCount += 1

        return pushesCount
